src/wh_client_crypto.c

```diff
@@ -156,6 +157,284 @@ int wh_Client_RngGenerate(whClientContext* ctx, uint8_t* out, uint32_t size)
     return ret;
 }
 
+#ifndef NO_AES
+
+#ifdef HAVE_AES_CBC
+int wh_Client_AesCbc(whClientContext* ctx,
+        Aes* aes, int enc,
+        uint32_t len, const uint8_t* in,
+        uint8_t* out)
+{
+    int ret = 0;
+    uint16_t group      = WH_MESSAGE_GROUP_CRYPTO;
+    uint16_t action     = WC_ALGO_TYPE_CIPHER;
+    uint16_t type       = WC_CIPHER_AES_CBC;
+    uint16_t dataSz     = 0;
+
+    uint16_t blocks = len / AES_BLOCK_SIZE;
+    uint32_t last_offset = (blocks - 1) * AES_BLOCK_SIZE;
+
+
+    uint32_t key_len    = aes->keylen;
+    const uint8_t* key  = (const uint8_t*)(aes->devKey);
+    whKeyId key_id      = WH_DEVCTX_TO_KEYID(aes->devCtx);
+
+    whPacket* packet    = (whPacket*)wh_CommClient_GetDataPtr(ctx->comm);
+
+    uint8_t* iv         = (uint8_t*)aes->reg;
+    uint32_t iv_len     = AES_IV_SIZE;
+
+    /* Request packet */
+    wh_Packet_cipher_aescbc_req* req = &packet->cipherAesCbcReq;
+    uint8_t* req_in     = (uint8_t*)(req + 1);
+    uint8_t* req_key    = req_in + len;
+    uint8_t* req_iv     = req_key + key_len;
+
+    uint32_t req_len    = WH_PACKET_STUB_SIZE + sizeof(*req) + len +
+                            key_len + iv_len;
+
+    /* Response packet */
+    wh_Packet_cipher_aescbc_res* res = &packet->cipherAesCbcRes;
+    uint8_t* res_out    = (uint8_t*)(res + 1);
+
+    uint32_t res_len    = WH_PACKET_STUB_SIZE+ sizeof(*res) + len;
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[client] %s: enc:%d keylen:%d ivsz:%d insz:%d reqsz:%u ressz:%u "
+                      "blocks:%u lastoffset:%u\n",
+                __func__,enc, key_len, iv_len, len,  req_len, res_len,
+                blocks, last_offset);
+    printf("[client] %s: req:%p in:%p key:%p iv:%p res:%p out:%p\n",
+            __func__, req, req_in, req_key, req_iv, res, res_out);
+#endif
+
+    /* III 0 size check is done in wolfCrypt */
+    if(     (blocks == 0) ||
+            ((len % AES_BLOCK_SIZE) != 0) ) {
+        /* CBC requires only full blocks */
+        return BAD_LENGTH_E;
+
+    }
+
+    if (    (req_len > WOLFHSM_CFG_COMM_DATA_LEN) ||
+            (res_len > WOLFHSM_CFG_COMM_DATA_LEN)) {
+        /* if we're using an HSM key return BAD_FUNC_ARG */
+        if (WH_KEYID_ISERASED(key_id)) {
+            return CRYPTOCB_UNAVAILABLE;
```

Comment:
Are we sure we want  to silently fall back to local crypto here? I guess I'm trying to figure out what would get you here....is this triggered when you want to do HSM crypto, have an input larger than the comm MTU, but but didn't call the Client SetKeyID API? 

Perhaps this should be `BAD_FUNC_ARG` too then? I feel like if they supply the HSM devID, then they should expect the operation to be RPC'ed to the HSM, not secretly done locally with no way of checking. And if it is too big to send across the transport, it should just be an error right?

That said, I think the key would almost always be NULL in this case since they most likely didn't call setKey if using an HSM key, so believe the regular AES logic would check it. Though if for some reason they called AesSetKey I think this would end up doing the operation locally without any error reported to the user, which I don't think we should be doing?

Reply:
If there is a key_id set, then you can't fallback and currently this function can't handle inputs/outputs that are larger than the MTU.  If there is no keyid set, then the key is local and we can fall back to local AES, if available.  This check is if the keyid is erased (not set) then assume local fallback is ok, otherwise the hsm client function was incorrectly called (badargs), which is a calling/coding error.